pyquil/api/_qpu.py

```diff
@@ -120,6 +121,13 @@ def run(self):
 
         :return: The QPU object itself.
         """
+        # This prevents a common error where users expect QVM.run()
+        # and QPU.run() to be interchangeable. QPU.run() needs the
+        # supplied executable to have been compiled, QVM.run() does not.
```

Comment:
This is good, I don't remember what the error would be otherwise, but this error message is certainly better.

Ideally, QPU and QVM _should be_ interchangeable.
I think a warning is raised when running on QVM without calling .compile, or maybe only in some cases. Maybe we can consider rewording this comment to encourage users to call .compile even when targeting the QVM, to prevent people from running into this issue in the first place.

Reply:
the full story here: https://github.com/rigetti/pyquil/pull/799#discussion_r254516188